src/dune_pkg/opam_solver.ml

```diff
@@ -230,6 +230,80 @@ module Summary = struct
   ;;
 end
 
+let opam_command_to_string_debug (args, _filter_opt) =
+  List.map args ~f:(fun (simple_arg, _filter_opt) ->
+    match simple_arg with
+    | OpamTypes.CString s -> s
+    | CIdent ident ->
+      (* Stringify the variable in the syntax of an opam string interpolation. *)
+      Printf.sprintf "%%{%s}%%" ident)
+  |> String.concat ~sep:" "
+;;
+
+let opam_commands_to_actions package (commands : OpamTypes.command list) =
+  let pform_opt_of_ident ident =
+    let `Self, variable =
+      match String.split ident ~on:':' with
+      | [ variable ] | [ "_"; variable ] -> `Self, variable
+      | _ ->
+        (* TODO *)
+        Code_error.raise
+          "Evaluating package variables for non-self packages not yet implemented"
+          [ "While processing package:", Dyn.string (OpamPackage.to_string package)
+          ; "Variable:", Dyn.string ident
+          ]
+    in
+    match Pform.Var.Pkg.opt_of_opam_variable_name variable with
+    | Some pkg_var -> Ok Pform.(Var (Var.Pkg pkg_var))
+    | None -> Error (`Unknown_variable variable)
+  in
+  List.filter_map commands ~f:(fun ((args, _filter_opt) as command) ->
+    let terms =
+      List.map args ~f:(fun (simple_arg, _filter_opt) ->
+        match simple_arg with
+        | OpamTypes.CString s ->
+          (* TODO: apply replace string interpolation variables with pforms *)
+          String_with_vars.make_text Loc.none s
+        | CIdent ident ->
+          (match pform_opt_of_ident ident with
+           | Ok pform -> String_with_vars.make_pform Loc.none pform
+           | Error (`Unknown_variable variable) ->
+             (* Note that the variable name is always quoted to clarify
+                the error message in cases where the grammar of the
+                sentence would otherwise be unclear, such as:
+
+                - Encountered unknown variable type while processing...
+                - Encountered unknown variable name while processing...
+
+                In these examples, the words "type" and "name" are variable
+                names but it would be easy for users to misunderstand those
+                error messages without quotes. *)
+             User_error.raise
+               [ Pp.textf
+                   "Encountered unknown variable \"%s\" while processing commands for \
+                    package %s."
+                   variable
+                   (OpamPackage.to_string package)
+               ; Pp.text "The full command:"
+               ; Pp.text (opam_command_to_string_debug command)
```

Comment:
I wonder whether it wouldn't be better to express this in the OPAM file syntax, because if I look at the test it states `%%{fake}%% install` but this doesn't occur in the source, instead it states `fake "install"` which inverts which parts get quoted and which ones don't.